File: TypeScript/735. Asteroid Collision.ts
// noinspection JSUnusedGlobalSymbols
const asteroidCollision = (asteroids: number[]): number[] => {
    const stack: number[] = []

    for (let i = 0; i < asteroids.length; i++) {
        const current = asteroids[i]

        if (current > 0) {
            stack.push(current)
        } else {
            const absCurrent = Math.abs(current)

            while (stack.length > 0 && stack[stack.length - 1] > 0 && stack[stack.length - 1] < absCurrent) {
                stack.pop()
            }

            const endStack = stack[stack.length - 1]
            const stackLength = stack.length

            if (stackLength > 0 && endStack === absCurrent) {
                stack.pop()
            } else if (stackLength === 0 || endStack < 0) {
                stack.push(current)
            }
        }
    }

    return stack
}

File: Go/735. Asteroid Collision.go
package main

//goland:noinspection GoUnusedFunction
func asteroidCollision(asteroids []int) []int {
	var stack []int

	for i := 0; i < len(asteroids); i++ {
		current := asteroids[i]

		if current > 0 {
			stack = append(stack, current)
		} else {
			absCurrent := -current

			for len(stack) > 0 && stack[len(stack)-1] > 0 && stack[len(stack)-1] < absCurrent {
				stack = stack[:len(stack)-1]
			}

			stackLength := len(stack)

			if stackLength > 0 && stack[stackLength-1] == absCurrent {
				stack = stack[:len(stack)-1]
			} else if stackLength == 0 || stack[stackLength-1] < 0 {
				stack = append(stack, current)
			}
		}
	}

	return stack
}
